crates/symbolicator-service/src/services/sourcemap.rs

```diff
@@ -1,338 +1,42 @@
-use super::fetch_file;
-use super::symbolication::JsProcessingSymbolicateStacktraces;
-
-pub type OwnedSourceMapCache = SelfCell<ByteView<'static>, SourceMapCache<'static>>;
-
-/// A URL to a sourcemap file.
-///
-/// May either be a conventional URL or a data URL containing the sourcemap
-/// encoded as BASE64.
-#[derive(Debug, Clone, PartialEq, Eq)]
-enum SourceMapUrl {
-    Data(Vec<u8>),
-    Remote(url::Url),
-}
-
-impl SourceMapUrl {
-    /// The string prefix denoting a data URL.
-    const DATA_PREAMBLE: &str = "data:application/json;base64,";
-
-    /// Parses a string into a [`SourceMapUrl`].
-    ///
-    /// If the string starts with [`DATA_PREAMBLE`](Self::DATA_PREAMBLE), the rest is decoded from BASE64.
-    /// Otherwise, the string is joined to the `base` URL.
-    fn parse_with_prefix(base: &Url, url_string: &str) -> CacheEntry<Self> {
-        if let Some(encoded) = url_string.strip_prefix(Self::DATA_PREAMBLE) {
-            let decoded = BASE64
-                .decode(encoded.as_bytes())
-                .map_err(|_| CacheError::Malformed("Invalid base64 sourcemap".to_string()))?;
-            Ok(Self::Data(decoded))
-        } else {
-            let url = base
-                .join(url_string)
-                .map_err(|_| CacheError::DownloadError("Invalid sourcemap url".to_string()))?;
-            Ok(Self::Remote(url))
-        }
-    }
-}
+use super::sourcemap_lookup::{
+    FetchArtifactCacheInternal, FetchSourceMapCacheInternal, SourceMapLookup,
+};
 
 #[derive(Debug, Clone)]
 pub struct SourceMapService {
+    artifact_caches: Arc<Cacher<FetchArtifactCacheInternal>>,
     sourcemap_caches: Arc<Cacher<FetchSourceMapCacheInternal>>,
     download_svc: Arc<DownloadService>,
 }
 
 impl SourceMapService {
     pub fn new(
-        cache: Cache,
+        artifact_cache: Cache,
+        sourcemap_cache: Cache,
         shared_cache: SharedCacheRef,
         download_svc: Arc<DownloadService>,
     ) -> Self {
         Self {
-            sourcemap_caches: Arc::new(Cacher::new(cache, shared_cache)),
+            artifact_caches: Arc::new(Cacher::new(artifact_cache, shared_cache.clone())),
+            sourcemap_caches: Arc::new(Cacher::new(sourcemap_cache, shared_cache)),
             download_svc,
         }
     }
 
-    pub async fn list_artifacts(
-        &self,
-        source: Arc<SentrySourceConfig>,
-    ) -> HashMap<String, SearchArtifactResult> {
-        self.download_svc
-            .list_artifacts(source)
-            .await
-            .into_iter()
-            .map(|artifact| (artifact.name.clone(), artifact))
-            .collect()
-    }
-
-    pub async fn fetch_artifact(
-        &self,
-        source: Arc<SentrySourceConfig>,
-        file_id: SentryFileId,
-    ) -> CacheEntry<ByteView<'static>> {
-        let mut temp_file = NamedTempFile::new().unwrap();
-        fetch_file(
+    pub fn create_sourcemap_lookup(&self, source: Arc<SentrySourceConfig>) -> SourceMapLookup {
```

Comment:
NOTE: I wanted to create a per-request (with internal `Scope` / `sources`) for native requests as well.